src/cargo/ops/cargo_generate_lockfile.rs

```diff
@@ -106,4 +121,23 @@ pub fn update_lockfile(manifest_path: &Path,
             None => {}
         }
     }
+
+    fn compare_dependency_graphs<'a>(previous_resolve: &'a Resolve,
+                                     resolve: &'a Resolve) ->
+                                     Vec<(Option<&'a PackageId>, Option<&'a PackageId>)> {
+        let mut changes = HashMap::new();
+        for dep in previous_resolve.iter() {
+            changes.insert(dep.name(), (Some(dep), None));
```

Comment:
I think that this may not be quite right as it's possible to have a few different situations:

* There could be multiple packages in the graph with the same name, but from different sources.
* There could be multiple versions of the same package in the graph, locked at different versions.

I think that this could be solved by having the key of the map being `(&str, &SourceId)` (e.g. `.name()` and `.source()`), and then the values would be `(Vec<&PackageId>, Vec<&PackageId>)` where the first element is all activated versions beforehand and the latter is all activated versions afterwards.

When the map is returned the left vector could represent "versions removed" and the right vector could represent "versions added". In the common case of updating a dep we'll remove one version and add one, but it could possibly be more complicated.